### PR TITLE
Lua preprocessor changes & editor support

### PR DESCRIPTION
**Pull request by matgis**
Implements #6303 in the editor and changes the pipeline plugin extension mechanism. This is a breaking change for Bob `LuaBuilderPlugin` implementors.

### User-facing changes
* The editor will now run any Lua preprocessors found in dependencies when building.

### Technical changes
* (Breaking Change) The `com.dynamo.bob.pipeline.LuaBuilderPlugin` abstract base class has been removed and broken up into two interfaces, `ILuaPreprocessor` and `ILuaObfuscator` in the new `com.defold.extension.pipeline` package.
* (Breaking Change) We now scan for Lua pipeline plugins in the `com.defold.extension.pipeline` package instead of the internal `com.dynamo.bob.pipeline` package.

### Migration instructions if we consider the `create` method of `LuaBuilderPlugin` to have been relased
Any class that inherited from the `com.dynamo.bob.pipeline.LuaBuilderPlugin` interface should now instead implement one, or both of the interfaces `ILuaPreprocessor` and `ILuaObfuscator` in the new `com.defold.extension.pipeline` package.
* If you implemented the `create` method, implement the `preprocess` method of the `ILuaPreprocessor` interface. Take note that the argument order has changed.
* If you implemented the `build` method, implement the `obfuscate` method of the `ILuaObfuscator` interface. Take note that you now get the file path and build variant as arguments in addition to the source code.

You must also move your implementation from the `com.dynamo.bob.pipeline` package to the new `com.defold.extension.pipeline` package. We will look for build pipeline extensions in this new package from now on.

#### Previous implementation
Both preprocessing and obfuscation must be handled in a single class.
```
package com.dynamo.bob.pipeline;

public class MyLuaTransformer extends LuaBuilderPlugin {
    @Override
    public String create(String filePath, String input, String buildVariant) throws Exception { ... }

    @Override
    public String build(String input) throws Exception { ... }
}
```

#### Updated implementation
Note that you can now split the implementation of preprocessing and obfuscation across two different classes if you so desire.
```
package com.defold.extension.pipeline;

public class MyLuaTransformer implements ILuaPreprocessor, ILuaObfuscator {
    @Override // Previously named "create". Note new argument order!
    public String preprocess(String source, String path, String buildVariant) throws Exception { ... }

    @Override // Previously named "build". Note added arguments!
    public String obfuscate(String source, String path, String buildVariant) throws Exception { ... }
}
```

## PR checklist (remove before merging)

* [x] Prepared the PR for release notes
	* [x] Proper description of feature or fix (see example x?)
	* [x] Added a #fixes ### if it fixes an issue
	* [x] Assigned issue to a project
* [ ] Added documentation for public API changes
* [ ] Is this a new feature?
	* [ ] Added engine and/or editor unit tests
	* [ ] Added or updated manual entry

### CODE REVIEW

File: editor/src/clj/editor/workspace.clj [hunk 1]
```diff
@@ -33,13 +34,21 @@ ordinary paths."
...
 
 (set! *warn-on-reflection* true)
 
+;; Class loader used when loading editor extensions from libraries.
+;; It's important to use the same class loader, so the type signatures match.
+(def ^:private ^DynamicClassLoader class-loader (DynamicClassLoader. (.getContextClassLoader (Thread/currentThread))))
+
+(defn load-class! [class-name]
+  (Class/forName class-name true class-loader))
+
```
Comment (matgis) [inline editor/src/clj/editor/workspace.clj:51]: Moved up in the file.

File: editor/src/clj/editor/workspace.clj [hunk 2]
```diff
@@ -402,12 +422,6 @@ ordinary paths."
     (when (string/includes? (resource/proj-path resource) "/plugins/bin/")
       (.setExecutable target-path true))))
 
-; It's important to use the same class loader, so that the type signatures match
-(def class-loader (clojure.lang.DynamicClassLoader. (.getContextClassLoader (Thread/currentThread))))
-
-(defn load-class! [class-name]
-  (Class/forName class-name true class-loader))
-
```
Comment (matgis) [inline editor/src/clj/editor/workspace.clj:410]: Moved up in the file.

File: editor/test/integration/extension_spine_test.clj [hunk 3]
```diff
@@ -12,7 +12,7 @@
 ;; CONDITIONS OF ANY KIND, either express or implied. See the License for the
 ;; specific language governing permissions and limitations under the License.
 
-(ns integration.spine-extension-test
+(ns integration.extension-spine-test
```
Comment (matgis) [inline editor/test/integration/extension_spine_test.clj:15]: Renamed this test for consistency so it corresponds to the name of `extension-spine`.

File: editor/test/integration/non_editable_resources_test.clj [hunk 4]
```diff
@@ -170,50 +170,50 @@
   (let [workspace (tu/setup-workspace! world project-path)
         project (tu/setup-project! workspace)]
     (tu/make-atlas-resource-node! project "/assets/from-script.atlas")
-    (tu/make-atlas-resource-node! project "/assets/from-chair-sprite.atlas")
-    (tu/make-atlas-resource-node! project "/assets/from-room-embedded-chair-sprite.atlas")
+    (tu/make-atlas-resource-node! project "/assets/from-chair-embedded-sprite.atlas")
+    (tu/make-atlas-resource-node! project "/assets/from-room-embedded-chair-embedded-sprite.atlas")
```
Comment (matgis) [inline editor/test/integration/non_editable_resources_test.clj:174]: Unrelated cleanup: Use consistent names for these assets across all tests.

File: editor/test/integration/script_properties_test.clj [hunk 5]
```diff
@@ -36,17 +36,6 @@
...
-  (assert (map? property-declarations))
-  (into {}
-        (mapcat (fn [[entries-key values-key]]
-                  (let [entries (get property-declarations entries-key)
-                        values (get property-declarations values-key)]
-                    (map (fn [{:keys [key index]}]
-                           [key (values index)])
-                         entries))))
-        (vals properties/type->entry-keys)))
-
```
Comment (matgis) [inline editor/test/integration/script_properties_test.clj:49]: I've moved a bunch of helper functions to the `test-util` module so I can use them across all the tests.

File: editor/test/integration/test_util.clj [hunk 6]
```diff
@@ -56,6 +60,8 @@
            [javax.imageio ImageIO]
            [org.apache.commons.io FilenameUtils IOUtils]))
 
+(set! *warn-on-reflection* true)
+
```
Comment (matgis) [inline editor/test/integration/test_util.clj:64]: Added reflection check and fixed issues below.

File: editor/test/integration/test_util.clj [hunk 7]
```diff
@@ -109,7 +115,12 @@
   (string/join "\n" (prop script-id :modified-lines)))
 
 (defn code-editor-source! [script-id source]
-  (prop! script-id :modified-lines (string/split source #"\r?\n" -1)))
+  (let [lines (cond
+                (string? source) (code.data/string->lines source)
+                (vector? source) source
+                :else (throw (ex-info "source must be a string or a vector of lines."
+                                      {:source source})))]
+    (prop! script-id :modified-lines lines)))
```
Comment (matgis) [inline editor/test/integration/test_util.clj:123]: Made this helper function accept both a string or a vector of lines.

File: editor/src/clj/internal/util.clj [hunk 8]
```diff
@@ -660,3 +660,15 @@
...
+  "Given a sequence of non-unique items, returns a function that will return the
+  same instance of each item if it is determined to be equal to an item already
+  in the ever-growing set of unique items."
+  [items]
+  (let [unique-volatile (volatile! (set items))]
+    (fn dedupe-fn [item]
+      (or (get @unique-volatile item)
+          (do
+            (vswap! unique-volatile conj item)
+            item)))))
```
Comment (matgis) [inline editor/src/clj/internal/util.clj:674]: We use this to split the string we get back from the Lua preprocessors into lines in an attempt to keep duplicate strings out of the cache since a lot of the lines in a preprocessed file will be kept the same.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java [hunk 9]
```diff
@@ -69,10 +69,6 @@ private static void scanJar(URL resource, String pkgname, Set<String> classes) t
         }
     }
 
-    protected URL resolveURL(URL url) throws IOException {
-        return url;
-    }
-
```
[the remark this answers is not in the document]
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java:75]: Probably so that it is easier to call from the editor? But we could keep both and call the static version from the non-static one?

File: editor/src/clj/internal/util.clj (hunk 8 above)
Comment (vlaaad) [reply editor/src/clj/internal/util.clj:674]: I don't understand how it works... Is it basically the same as `identity`? Just returns the input... Or is it for ensuring that for equal items we preserve only identical ones?

If it's for ensuring identical values, I'm not sure `get` from the set works:
```clj
(let [string-set #{(str "fo" "o")}
      s "foo"]
  {:equal (= (first string-set) s)
   :identical (identical? (first string-set) s)
   :get (identical? s (get string-set s))})
=> {:equal true
    :identical false
    :get false}
```

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ClassLoaderScanner.java:75]: it seems like we keep both here, I just don't get why... ok

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 10]
```diff
@@ -90,16 +94,21 @@ private LuaScanner getLuaScanner(IResource resource) throws IOException, Compile
...
-                        throw new CompileExceptionError(resource, 0, "Unable to run Lua builder plugin", e);
-                    }
+            // Create and run preprocessors if some exists.
+            if (luaPreprocessors == null) {
+                luaPreprocessors = PluginScanner.getOrCreatePlugins("com.defold.extension.pipeline", ILuaPreprocessor.class);
+
+                if (luaPreprocessors == null) {
+                    luaPreprocessors = new ArrayList<ILuaPreprocessor>(0);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (AGulev) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:104]: `PluginScanner.getOrCreatePlugins` has its own cache inside which makes it not necessary

**Review by vlaaad — COMMENTED** [review]
LGTM, but I'm very confused about the dedupe-fn...

File: editor/src/clj/internal/util.clj (hunk 8 above)
Comment (matgis) [reply editor/src/clj/internal/util.clj:674]: It's intended to make sure you get a reference to the original item when it is equal to the value you put in. I think your `:get` check is wrong? It should be:
```clj
(let [string-set #{(str "fo" "o")}
      s "foo"]
  {:equal (= (first string-set) s)
   :identical (identical? (first string-set) s)
   :get (identical? (first string-set) (get string-set s))})
=> {:equal true,
    :identical false,
    :get true}
```
I wrote a unit test for it in `internal/util_test.clj` that you can see further down in the pull request.

File: editor/src/clj/editor/code/data.clj [hunk 11]
```diff
@@ -336,6 +336,14 @@
 (defn lines-input-stream ^InputStream [lines]
   (ReaderInputStream. (lines-reader lines)))
 
+(defn lines->string
+  ^String [lines]
+  (with-open [reader (lines-reader lines)]
+    (slurp reader)))
```
[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/code/data.clj:342]: Yeah, maybe. `slurp` uses a `StringWriter` which does its own buffering internally so it should be reasonably efficient like this. I feel we can merge this as is for now, but I'd like to look into it at some point.

**Review by AGulev — DISMISSED** [review]

File: editor/src/clj/internal/util.clj (hunk 8 above)
Comment (vlaaad) [reply editor/src/clj/internal/util.clj:674]: Yes, my mistake, that makes sense!

File: editor/src/clj/editor/code/data.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (vlaaad) [reply editor/src/clj/editor/code/data.clj:342]: But it's 5 times faster!
```clj
(let [lines (editor.code.util/split-lines (slurp "doc/defnode.md"))]
  (println "slurp")
  (criterium.core/quick-bench
    (with-open [r (editor.code.data/lines-reader lines)]
      (slurp r)))
  (println "string builder")
  (criterium.core/quick-bench
    (transduce
      (interpose "\n")
      (fn
        ([] (StringBuilder.))
        ([ret] (str ret))
        ([^StringBuilder acc ^String input] (doto acc (.append input))))
      lines)))
;slurp
;WARNING: Final GC required 5.102048233430867 % of runtime
;Evaluation count : 1896 in 6 samples of 316 calls.
;             Execution time mean : 349,955425 µs
;    Execution time std-deviation : 23,562550 µs
;   Execution time lower quantile : 318,034538 µs ( 2,5%)
;   Execution time upper quantile : 377,087725 µs (97,5%)
;                   Overhead used : 6,999291 ns
;
;Found 1 outliers in 6 samples (16,6667 %)
;    low-severe     1 (16,6667 %)
; Variance from outliers : 15,0710 % Variance is moderately inflated by outliers
;string builder
;WARNING: Final GC required 38.470103234406935 % of runtime
;Evaluation count : 9900 in 6 samples of 1650 calls.
;             Execution time mean : 62,009516 µs
;    Execution time std-deviation : 3,253784 µs
;   Execution time lower quantile : 60,132496 µs ( 2,5%)
;   Execution time upper quantile : 67,606398 µs (97,5%)
;                   Overhead used : 6,999291 ns
;
;Found 1 outliers in 6 samples (16,6667 %)
;    low-severe     1 (16,6667 %)
; Variance from outliers : 14,0394 % Variance is moderately inflated by outliers
```

File: editor/src/clj/editor/code/preprocessors.clj [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                    (when (java/public-implementation? uninitialized-class ILuaPreprocessor)
+                      (let [initialized-class (Class/forName class-name true class-loader)]
+                        {:type :success
+                         :value initialized-class})))
+                  (catch Exception e
+                    (log/error :msg (str "Exception in static initializer of ILuaPreprocessor implementation " class-name ": " (.getMessage e))
+                               :exception e)
+                    {:type :error
+                     :class-name class-name}))))
+        (ClassLoaderScanner/scanClassLoader class-loader scanned-package-name)))
```
[the remark this answers is not in the document]
Comment (vlaaad) [reply editor/src/clj/editor/code/preprocessors.clj:107]: Yes, indeed I missed it 😅 

Regarding `group-by` (or even better, `group-into`), this function produces a flat list of successes and failures, but the users of this function then filter it twice, first for failures (`(filter #(= :error (:type %)) initialize-infos)`), then for successes (`(success-values initialize-infos)`). You could instead return a map with errors and successes produced with `group-into`, saving time on extra filtering.

File: editor/src/clj/editor/code/data.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/code/data.clj:342]: I guess I'll make the change! 😅

File: editor/src/clj/editor/code/preprocessors.clj (hunk 12 above)
[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/code/preprocessors.clj:107]: Makes sense. I'll make the change!

File: editor/src/clj/editor/code/preprocessors.clj [hunk 13]
```diff
@@ -144,20 +143,18 @@ return m
...
     (when (not= old-lua-preprocessor-classes new-lua-preprocessor-classes)
-      (let [create-infos (try-create-lua-preprocessors lua-preprocessor-classes)
-            errors (filter #(= :error (:type %)) create-infos)]
-        (if (seq errors)
-          (report-error! "Failed to construct Lua preprocessors" (map :class-name errors))
-          (let [new-lua-preprocessors (success-values create-infos)]
-            (g/set-property! code-preprocessors :lua-preprocessors new-lua-preprocessors)))))))
+      (let [{:keys [created-lua-preprocessors faulty-class-names]}
+            (try-create-lua-preprocessors lua-preprocessor-classes)]
+        (if (seq faulty-class-names)
```
Comment (vlaaad) [inline editor/src/clj/editor/code/preprocessors.clj:148]: Nitpick: I think you can use `faulty-class-names` instead of `(seq faulty-class-names)` here — because of group-into, this key will be either absent or contain a non-empty vector.

**Review by vlaaad — APPROVED** [review]
Ship it!
